src/index.js

```diff
@@ -0,0 +1,269 @@
+    }
+    ++existing._references
+
+    return existing
+  }
+
+  _removePeer (peer) {
+    const id = peer.info.id.toB58String()
+
+    this.log('remove', id, peer._references)
+    // Only delete when no one else is referencing this peer.
+    if (--peer._references === 0) {
+      this.log('delete peer', id)
+      this.peers.delete(id)
+    }
+
+    return peer
+  }
+
+  _dialPeer (peerInfo, callback) {
+    callback = callback || function noop () { }
+    const idB58Str = peerInfo.id.toB58String()
+
+    // If already have a PubSub conn, ignore
+    const peer = this.peers.get(idB58Str)
+    if (peer && peer.isConnected) {
+      return nextTick(() => callback())
+    }
+
+    // If already dialing this peer, ignore
+    if (this._dials.has(idB58Str)) {
+      this.log('already dialing %s, ignoring dial attempt', idB58Str)
+      return nextTick(() => callback())
+    }
+    this._dials.add(idB58Str)
+
+    this.log('dialing %s', idB58Str)
+    this.libp2p.dialProtocol(peerInfo, this.multicodec, (err, conn) => {
+      this.log('dial to %s complete', idB58Str)
+
+      // If the dial is not in the set, it means that pubsub has been
+      // stopped
+      const pubsubStopped = !this._dials.has(idB58Str)
+      this._dials.delete(idB58Str)
+
+      if (err) {
+        this.log.err(err)
+        return callback()
+      }
+
+      // pubsub has been stopped, so we should just bail out
+      if (pubsubStopped) {
+        this.log('pubsub was stopped, not processing dial to %s', idB58Str)
+        return callback()
+      }
+
+      this._onDial(peerInfo, conn, callback)
+    })
+  }
+
+  _onDial (peerInfo, conn, callback) {
+    const idB58Str = peerInfo.id.toB58String()
+    this.log('connected', idB58Str)
+
+    const peer = this._addPeer(new Peer(peerInfo))
+    peer.attachConnection(conn)
+
+    nextTick(() => callback())
+  }
+
+  _onConnection (protocol, conn) {
+    conn.getPeerInfo((err, peerInfo) => {
+      if (err) {
+        this.log.err('Failed to identify incomming conn', err)
+        return pull(empty(), conn)
+      }
+
+      const idB58Str = peerInfo.id.toB58String()
+      const peer = this._addPeer(new Peer(peerInfo))
+
+      this._processConnection(idB58Str, conn, peer)
+    })
+  }
+
+  /**
+   * Process connections
+   * @abstract
+   * @param {string} idB58Str peer id string in base58
+   * @param {Connection} conn connection
+   * @param {PeerInfo} peer peer info
+   * @returns {undefined}
+   *
+   */
+  _processConnection (idB58Str, conn, peer) {
+    throw errcode('_processConnection must be implemented by the subclass', 'ERR_NOT_IMPLEMENTED')
+  }
+
+  _onConnectionEnd (idB58Str, peer, err) {
+    // socket hang up, means the one side canceled
+    if (err && err.message !== 'socket hang up') {
+      this.log.err(err)
+    }
+
+    this.log('connection ended', idB58Str, err ? err.message : '')
+    this._removePeer(peer)
+  }
+
+  /**
+   * Publish messages to the given topics.
```

Comment:
Similar to the processConnection comment. Publish is a clearer term in regards to pubsub, but I think it would still be helpful to spell out expectations. Something like `Overriding implementations of publish should handle the appropriate algorithms for their implementations. For example, a Floodsub implementation might simply publish each message to each topic for every peer, whereas a Gossipsub implementation may be more selective about the peers it publishes to.`